tests/functional/pyocf/types/volume.py

```diff
@@ -66,75 +66,132 @@ class VolumeProperties(Structure):
         ("_caps", VolumeCaps),
         ("_io_ops", IoOps),
         ("_deinit", c_char_p),
-        ("_ops", VolumeOps),
+        ("_ops_", VolumeOps),
     ]
 
-
 class VolumeIoPriv(Structure):
     _fields_ = [("_data", c_void_p), ("_offset", c_uint64)]
 
 
-class Volume(Structure):
-    VOLUME_POISON = 0x13
+VOLUME_POISON = 0x13
+
 
-    _fields_ = [("_storage", c_void_p)]
+class Volume:
     _instances_ = weakref.WeakValueDictionary()
     _uuid_ = weakref.WeakValueDictionary()
+    _ops_ = {}
+    _props_ = {}
 
-    props = None
+    @classmethod
+    def get_ops(cls):
+        if cls in Volume._ops_:
+            return Volume._ops_[cls]
+
+        @VolumeOps.SUBMIT_IO
+        def _submit_io(io):
+            io_structure = cast(io, POINTER(Io))
+            volume = Volume.get_instance(
+                OcfLib.getInstance().ocf_io_get_volume(io_structure)
+            )
 
-    def __init__(self, size: S, uuid=None):
-        super().__init__()
-        self.size = size
-        if uuid:
-            if uuid in type(self)._uuid_:
-                raise Exception(
-                    "Volume with uuid {} already created".format(uuid)
-                )
-            self.uuid = uuid
-        else:
-            self.uuid = str(id(self))
+            volume.submit_io(io_structure)
 
-        type(self)._uuid_[self.uuid] = self
+        @VolumeOps.SUBMIT_FLUSH
+        def _submit_flush(flush):
+            io_structure = cast(flush, POINTER(Io))
+            volume = Volume.get_instance(
+                OcfLib.getInstance().ocf_io_get_volume(io_structure)
+            )
 
-        self.data = create_string_buffer(int(self.size))
-        memset(self.data, self.VOLUME_POISON, self.size)
-        self._storage = cast(self.data, c_void_p)
+            volume.submit_flush(io_structure)
 
-        self.reset_stats()
-        self.opened = False
+        @VolumeOps.SUBMIT_METADATA
+        def _submit_metadata(meta):
+            raise NotImplementedError
 
-    def get_copy(self):
-        new_volume = Volume(self.size)
-        memmove(new_volume.data, self.data, self.size)
-        return new_volume
+        @VolumeOps.SUBMIT_DISCARD
+        def _submit_discard(discard):
+            io_structure = cast(discard, POINTER(Io))
+            volume = Volume.get_instance(
+                OcfLib.getInstance().ocf_io_get_volume(io_structure)
+            )
+
+            volume.submit_discard(io_structure)
+
+        @VolumeOps.SUBMIT_WRITE_ZEROES
+        def _submit_write_zeroes(write_zeroes):
+            raise NotImplementedError
+
+        @CFUNCTYPE(c_int, c_void_p)
```

Comment:
Why is this the only method, which is decorated generically and not with dedicated decorator like `VolumpeOps.OPEN`?

Reply:
fixed